redis/src/rustic_ai/redis/messaging/message_store.py

```diff
@@ -0,0 +1,252 @@
+        message_json = message.to_json()
+        msg_key = self._get_msg_key(namespace, message.id)
+
+        # Execute with retry logic
+        def store_operation():
+            # Use Redis transaction (MULTI/EXEC) for atomic storage
+            with self.client.pipeline(transaction=True) as pipe:
+                # All commands execute atomically when pipe.execute() is called
+                pipe.set(msg_key, message_json, ex=self.message_ttl)
+                # Use timestamp as score like the original implementation
+                pipe.zadd(topic, {message_json: message.timestamp})
+                pipe.execute()
+
+            # Publish separately (not part of transaction)
+            subscribers_notified = pubsub_publish(topic, message_json)
+            logging.debug(
+                f"Message {message.id} stored atomically for topic {topic}, notified {subscribers_notified} subscribers"
+            )
+
+        try:
+            self._execute_with_retry(f"Store message {message.id} to topic {topic}", store_operation)
+        except Exception as e:
+            logging.error(f"Atomic message storage failed for message {message.id} on topic {topic}: {e}")
+            raise
+
+    def get_messages_for_topic(self, topic: str) -> List[Message]:
+        """
+        Retrieve all messages for a given topic.
+
+        Args:
+            topic: The topic to retrieve messages for.
+
+        Returns:
+            List[Message]: A list of messages for the given topic.
+        """
+
+        def get_operation():
+            raw_messages = self.client.zrange(topic, 0, -1)
+
+            if isinstance(raw_messages, Awaitable):  # runtime check
+                raise RuntimeError("Unexpected awaitable from synchronous Redis client")
+
+            messages = [Message.from_json(raw_message) for raw_message in raw_messages]
+            # Sort by message.id like the original implementation for proper priority+timestamp ordering
+            return sorted(messages, key=lambda msg: msg.id)
+
+        return self._execute_with_retry(f"Get messages for topic {topic}", get_operation)
+
+    def get_messages_for_topic_since(self, topic: str, msg_id_since: int) -> List[Message]:
+        """
+        Retrieve all messages for a given topic since a given message ID.
+
+        Args:
+            topic: The topic to retrieve messages for.
+            msg_id_since: The ID of the message since which to retrieve messages.
+
+        Returns:
+            List[Message]: A list of messages for the given topic since the given message ID.
+        """
+
+        def get_operation():
+            # Use timestamp-based filtering since "since ID" means "since that message's timestamp"
+            # Higher priority messages created later can have smaller IDs, so ID filtering doesn't work
+            timestamp_since = self._get_timestamp_for_id(msg_id_since) + 1
+            raw_messages = self.client.zrangebyscore(topic, timestamp_since, "+inf")
```

Comment:
This command is deprecated - https://redis.io/docs/latest/commands/zrangebyscore/